src/server/lock_manager/mod.rs

```diff
@@ -225,78 +220,75 @@ impl LockManager {
-        let mut detected = self.detected[detected_slot_idx(txn_ts)].lock();
-        detected.insert(txn_ts);
-    }
-
-    fn remove_from_detected(&self, txn_ts: TimeStamp) -> bool {
-        let mut detected = self.detected[detected_slot_idx(txn_ts)].lock();
-        detected.remove(&txn_ts)
-    }
 }
 
 impl LockManagerTrait for LockManager {
+    fn allocate_token(&self) -> LockWaitToken {
+        LockWaitToken(Some(self.token_allocator.fetch_add(1, Ordering::Relaxed)))
+    }
+
     fn wait_for(
         &self,
+        token: LockWaitToken,
+        region_id: u64,
+        region_epoch: RegionEpoch,
+        term: u64,
         start_ts: TimeStamp,
-        cb: StorageCallback,
-        pr: ProcessResult,
-        lock: Lock,
+        wait_info: KeyLockWaitInfo,
         is_first_lock: bool,
         timeout: Option<WaitTimeout>,
+        cancel_callback: Box<dyn FnOnce(StorageError) + Send>,
         diag_ctx: DiagnosticContext,
     ) {
         let timeout = match timeout {
             Some(t) => t,
             None => {
-                cb.execute(pr);
+                Waiter::cancel_no_timeout(wait_info, cancel_callback);
                 return;
             }
         };
 
         // Increase `waiter_count` here to prevent there is an on-the-fly WaitFor msg
         // but the waiter_mgr haven't processed it, subsequent WakeUp msgs may be lost.
         self.waiter_count.fetch_add(1, Ordering::SeqCst);
-        self.waiter_mgr_scheduler
-            .wait_for(start_ts, cb, pr, lock, timeout, diag_ctx.clone());
 
         // If it is the first lock the transaction tries to lock, it won't cause
         // deadlock.
         if !is_first_lock {
-            self.add_to_detected(start_ts);
-            self.detector_scheduler.detect(start_ts, lock, diag_ctx);
+            self.detector_scheduler
+                .detect(start_ts, wait_info.clone(), diag_ctx.clone()); // TODO: Try to avoid cloning.
         }
+        self.waiter_mgr_scheduler.wait_for(
+            token,
+            region_id,
+            region_epoch,
+            term,
+            start_ts,
+            wait_info,
+            timeout,
+            cancel_callback,
+            diag_ctx,
+        );
     }
 
-    fn wake_up(
-        &self,
-        lock_ts: TimeStamp,
-        hashes: Vec<u64>,
-        commit_ts: TimeStamp,
-        is_pessimistic_txn: bool,
-    ) {
-        // If `hashes` is some, there may be some waiters waiting for these locks.
-        // Try to wake up them.
-        if !hashes.is_empty() && self.has_waiter() {
-            self.waiter_mgr_scheduler
-                .wake_up(lock_ts, hashes, commit_ts);
-        }
-        // If a pessimistic transaction is committed or rolled back and it once sent
-        // requests to detect deadlock, clean up its wait-for entries in the
-        // deadlock detector.
-        if is_pessimistic_txn && self.remove_from_detected(lock_ts) {
-            self.detector_scheduler.clean_up(lock_ts);
```

Comment:
Dumb question: is there an equivalent part of the clean up logic now?

Reply:
Here the old logic wakes up all waiters waiting for the lock with `lock_ts` on the specified key, and cleans all edges that waits for the `lock_ts` from the detector. In new logic, waking-up happens in the new lock waiting queue, and when a lock-waiting request is finished (either canceled or resumed and successfully acquired, which is not yet supported), the `remove_lock_wait` function in LockManager will be invoked, which then leads to a call to `clean_up_wait_for` ([here](https://github.com/tikv/tikv/blob/4785619d18027001c16af878fad567c02ae0dd42/src/server/lock_manager/waiter_manager.rs#L558))